WT-12713 Generate set_oldest_timestamp operations

pmacko86 opened the pull request:
This PR:
* Adds the oldest timestamp to `model::kv_database`.
* Adds the `set_oldest_timestamp` workload operation.
* Adds the corresponding support to the workload generator.
* Adds timestamp verification to tests and to the debug log verifier.
* Fixes a bug in the debug log parser, so that it can handle empty prepared transactions.
* Fixes a bug in the debug log based verifier, so that it better handles the previous key-value pair when printing an error.

This ended up being a surprisingly substantial PR; feel free to let me know if you'd like a walk-through.

korteland commented on test/model/src/include/model/kv_database.h:
```diff
@@ -101,6 +105,27 @@ class kv_database {
+    set_oldest_timestamp(timestamp_t timestamp) noexcept
+    {
+        std::lock_guard lock_guard(_timestamps_lock);
+        _oldest_timestamp = std::max(_oldest_timestamp, timestamp);
```
I'm not 100% sure that this matches WT behaviour - isn't it an error to try to make the oldest timestamp go backwards?

etienneptl replied:
It is an error, as described [here](https://github.com/wiredtiger/wiredtiger/pull/10427/files#diff-ba4c9e2fbaeaefdd722d5c7566e096e249c1ad3b59214e6b7443061ea9e9be2eR223).
Taking the max here means we should never have a lower value than the current `_oldest_timestamp`, it should be fine.

pmacko86 replied:
When trying this empirically, this is the same behavior as I get with WT, and I added unit tests to make sure.

It is not currently an error for the oldest timestamp to be set to an earlier value, just like with the stable timestamp (at least until Etienne's PR gets merged in). I still didn't want to generate that behavior in the workload generator, which is why I added the check.

korteland reviewed (commented):
Mostly LGTM - the details largely look fine but a quick chat to see how it hangs together overall and get the bigger picture would be useful to me.

etienneptl commented on test/model/src/core/kv_database.cpp:
```diff
@@ -102,8 +102,13 @@ kv_database::create_checkpoint_nolock(
             throw model_exception("Checkpoint already exists: " + ckpt_name);
     }
 
+    /* Remember the oldest timestamp only if the stable timestamp is set. */
```
That's an interesting case, is that what WT does as well?

pmacko86 replied:
Yes. I don't know if it's documented, but empirically I see that this is what happens. You can check that the unit tests verify that WT has the same behavior. But I agree that it is somewhat unexpected!

etienneptl replied:
It has been confirmed that this is how WT works. Whether it is the right behaviour or not, it is being discussed within the team. The outcome is beyond the scope of this PR - resolving.

etienneptl commented on test/model/src/driver/debug_log_parser.cpp:
```diff
@@ -556,8 +560,16 @@ debug_log_parser::commit_transaction(kv_transaction_ptr txn)
-    if (txn_state != kv_transaction_state::committed)
-        txn->commit();
+    if (txn_state != kv_transaction_state::committed) {
+        /*
```
np: This comment might be unexpected after the comment line 562. Maybe we could merge them both line 562?

pmacko86 replied:
I rewrote the comment in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

etienneptl commented on test/model/src/driver/kv_workload_generator.cpp:
```diff
@@ -215,6 +216,23 @@ kv_workload_generator::assert_timestamps(
         stable = t;
     }
 
+    if (std::holds_alternative<operation::set_oldest_timestamp>(op)) {
```
I have a question regarding lines above which manage the stable ts.
What happens in this scenario:

- Start WT, not ts are set
- Set oldest to 10
- Set stable to 9

The above will not trigger as we set stable for the first time, but it is lower than the oldest ts.

pmacko86 replied:
I also added a similar check to the stable timestamp branch above in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

etienneptl commented on test/model/test/common/wiredtiger_util.cpp:
```diff
@@ -86,13 +86,16 @@ wt_get_ext(WT_SESSION *session, const char *uri, const model::data_value &key,
+        ret = session->begin_transaction(session, cfg);
+        if (ret == EINVAL)
+            return ret;
+        testutil_check(ret);
```
Same question, shouldn't we fail for an invalid ts?

pmacko86 replied:
I need this for the unit tests that check that the model has the same behavior as WT when using a too-old read timestamp. Without this, doing the test on WT would result in an abort, but now I can just compare the return values.

korteland replied:
Got it - thanks.

etienneptl commented on test/model/test/common/wiredtiger_util.cpp:
```diff
@@ -422,6 +425,35 @@ wt_ckpt_create(WT_SESSION *session, const char *ckpt_name)
     testutil_check(session->checkpoint(session, config.c_str()));
 }
 
+/*
```
Can we create private helper functions for `wt_[get|set]_[stable|oldest]_timestamp`? For example:
```
wt_set_timestamp(WT_CONNECTION *conn, const std::string& timestamp_name, model::timestamp_t timestamp)
{
    char buf[64];

    testutil_snprintf(buf, sizeof(buf), "%s=%" PRIx64, timestamp_name, timestamp);
    testutil_check(conn->set_timestamp(conn, buf));
}
```
What do you think?

pmacko86 replied:
Good idea—fixed in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

etienneptl reviewed (commented):
Added a few comments, I will need another review.

korteland reviewed (approved):
LGTM! Thanks for the walkthrough.

pmacko86 commented:
In addition to addressing the PR comments, I also discovered and fixed a bug in which the workload generator would not properly keep track of the oldest and stable timestamps across database crashes, which would result in generating workloads like this:

- Set stable timestamp to 5
- Checkpoint
- Set oldest timestamp to 10
- Set stable timestamp to 20
- Crash
- Set oldest timestamp to 15 (thinking that the stable timestamp is still 20)

This is also included in [cc5b2fb](https://github.com/wiredtiger/wiredtiger/pull/10427/commits/cc5b2fb1f8ed4c230e66945e028a36edeca3fab4).

etienneptl reviewed (approved):
LGTM!
